firebase-database/src/main/java/com/google/firebase/database/FirebaseDatabase.java

```diff
@@ -302,9 +294,23 @@ public synchronized void setPersistenceCacheSizeBytes(long cacheSizeInBytes) {
     this.config.setPersistenceCacheSizeBytes(cacheSizeInBytes);
   }
 
-  @Nullable
-  private static EmulatedServiceSettings getEmulatorServiceSettings(@NonNull FirebaseApp app) {
-    return app.getEmulatorSettings().getServiceSettings(EMULATOR);
+  /**
+   * Modify this FirebaseDatabase instance to communicate with the Realtime Database emulator.
+   *
+   * <p>Note: this must be called before this instance has been used to do any database operations.
+   *
+   * @param host the emulator host (ex: 10.0.2.2)
+   * @param port the emulator port (ex: 9000)
+   */
+  public void useEmulator(@NonNull String host, int port) {
+    if (this.repo != null) {
+      throw new IllegalStateException(
+          "Cannot call useEmulator() after instance has already been initialized.");
+    }
+
+    getApp()
+        .get(FirebaseDatabaseComponent.class)
+        .setEmulatorSettings(new EmulatedServiceSettings(host, port));
```

Comment:
Why do we need to round-trip through Firebase App? It seems like this could now stay within the SDK.

I am also not sure if I understand the merits of proposal 2. We now have two places next to each other which both configure settings that can end up conflicting. Isn't the thing that is actually missing from our settings object just an "isEmulator" property? Or even - is this something the Emulator could just tell us?

Reply:
RTDB has no settings object ... did you mean for Firestore?  

I am at the end of my wits here I truly don't care how we design this.  I only did Proposal #2 because I got told off after proposal #1 was already implemented and merged so I reverted it.  I can't justify this API because it's not mine.

The reason this has to be app-scoped is because if you call `FirebaseDatabase.getInstance(app, url)` we need to apply the same emulator settings.

Reply:
I think I meant the url parameter in `FirebaseDatabase.getInstance(app, url)`, which is RTDB's "Settings" object. I will review this PR for what it is later today (looks fine glancing over it without too much depth).

Reply:
No more going through app.